### PR TITLE
Issue #2283 Fix special char handling in proxy

### PR DESCRIPTION
**Pull request by praveenkumar**
Fix : #2283

### CODE REVIEW

File: pkg/util/proxy.go [hunk 1]
```diff
@@ -161,3 +162,15 @@ func ValidateProxyURL(proxyUrl string) error {
...
 	return nil
 }
+
+// parseProxySpecialChar parse the URI and convert special char to hex
+func parseProxySpecialChar(uri string) string {
+	u, _ := url.Parse(uri)
+	scheme := fmt.Sprintf("%s://", u.Scheme)
+	if strings.HasPrefix(uri, scheme) {
+		s := strings.Replace(uri, scheme, "", 1)
+		s = url.PathEscape(s)
```
Comment (anjannath) [inline pkg/util/proxy.go]: :+1:

Comment (LalatenduMohanty) [reply pkg/util/proxy.go]: 👍

File: pkg/util/proxy_test.go [hunk 2]
```diff
@@ -162,3 +162,19 @@ func Test_http_proxy_from_env_lowercase_precedence(t *testing.T) {
...
+		{"", ""},
+		{"http://foo.com:3128", "http://foo.com:3128"},
+		{"http://user:F@oo!B#ar$@myserver:3128", "http://user:F@oo%21B%23ar$@myserver:3128"},
+		{"https://myuser:my#pass@foo.com:3128", "https://myuser:my%23pass@foo.com:3128"},
+	}
+	for _, proxyUrl := range urlList {
+		got := parseProxySpecialChar(proxyUrl.givenURI)
+		assert.Equal(t, proxyUrl.expectedURI, got)
+	}
+}
```
Comment (LalatenduMohanty) [inline pkg/util/proxy_test.go]: 👍

File: pkg/util/proxy.go (hunk 1 above)
Comment (LalatenduMohanty) [reply pkg/util/proxy.go]: If we can get confirmation that this works for the person who created the issue then we can merge it.

File: pkg/util/proxy_test.go (hunk 2 above)
Comment (amitkrout) [reply pkg/util/proxy_test.go]: To run the unit test on my local setup i am following the the upstream doc [1]. I am not sure what would be the ```<Testcase Name>``` for the command i wand to execute a single test on a single package
```
go test -v pkg/util/proxy_test.go -run <Testcase Name>
```
[1] https://docs.openshift.org/latest/minishift/contributing/developing.html#unit-tests

**Comment by anjannath** [comment]
@amit I think it will be the function name, can you please try with the
function name.

**Comment by amitkrout** [comment]
> I think it will be the function name, can you please try with the function name.

Too many errors
```
$ go test -v pkg/util/proxy_test.go -run Test_parse_special_character_uri
# command-line-arguments
pkg/util/proxy_test.go:27:22: undefined: NewProxyConfig
pkg/util/proxy_test.go:30:41: undefined: OpenShiftRegistryIp
pkg/util/proxy_test.go:34:22: undefined: NewProxyConfig
pkg/util/proxy_test.go:51:22: undefined: NewProxyConfig
pkg/util/proxy_test.go:67:12: undefined: NewProxyConfig
pkg/util/proxy_test.go:76:12: undefined: NewProxyConfig
pkg/util/proxy_test.go:85:22: undefined: NewProxyConfig
pkg/util/proxy_test.go:111:10: undefined: ValidateProxyURL
pkg/util/proxy_test.go:141:23: undefined: NewProxyConfig
pkg/util/proxy_test.go:159:22: undefined: NewProxyConfig
pkg/util/proxy_test.go:159:22: too many errors
FAIL    command-line-arguments [build failed]
```

File: pkg/util/proxy.go (hunk 1 above)
Comment (budhrg) [reply pkg/util/proxy.go]: Have we pointed the artifact binary to the user who created this issue? @praveenkumar @anjannath

**Comment by budhrg** [comment]
retest this please

File: pkg/util/proxy_test.go (hunk 2 above)
Comment (LalatenduMohanty) [reply pkg/util/proxy_test.go]: @amitkrout here is the right way to run it. You need to point to the the package directory i.e. `util`. 

```
$ go test -v ./pkg/util -run Test_parse_special_character_uri
=== RUN   Test_parse_special_character_uri
--- PASS: Test_parse_special_character_uri (0.00s)
PASS
ok  	github.com/minishift/minishift/pkg/util	0.017s
```

**Comment by LalatenduMohanty** [comment]
> Too many errors



> Too many errors

 Refer https://github.com/minishift/minishift/pull/2329#discussion_r183860384
```
$ go test -v ./pkg/util -run Test_parse_special_character_uri
=== RUN   Test_parse_special_character_uri
--- PASS: Test_parse_special_character_uri (0.00s)
PASS
ok  	github.com/minishift/minishift/pkg/util	0.017s
```

**Comment by praveenkumar** [comment]
retest this please.

File: pkg/util/proxy_test.go (hunk 2 above)
Comment (budhrg) [reply pkg/util/proxy_test.go]: Passing for me as well.

File: pkg/util/proxy_test.go [hunk 3]
```diff
@@ -162,3 +162,19 @@ func Test_http_proxy_from_env_lowercase_precedence(t *testing.T) {
...
+
+func Test_parse_special_character_uri(t *testing.T) {
+	var urlList = []struct {
+		givenURI    string
+		expectedURI string
+	}{
+		{"", ""},
+		{"http://foo.com:3128", "http://foo.com:3128"},
+		{"http://user:F@oo!B#ar$@myserver:3128", "http://user:F@oo%21B%23ar$@myserver:3128"},
+		{"https://myuser:my#pass@foo.com:3128", "https://myuser:my%23pass@foo.com:3128"},
```
Comment (amitkrout) [inline pkg/util/proxy_test.go]: @praveenkumar Can we add a new urlList for example ?
```
{"https://newuser:new(pas*)w^rd@test.com:3128", "https://newuser:new%28pas%2A%29w%5Erd@test.com:3128"}
```
Actually special character like (, *,^,) are missing in the password filed in the above list.

**Comment by budhrg** [comment]
Based on feedback from user who reported this issue https://github.com/minishift/minishift/issues/2283#issuecomment-384810256, I am merging this PR. 
Looks like PR is working.

**Pull request merged**